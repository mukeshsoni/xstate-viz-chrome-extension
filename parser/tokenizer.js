function last(arr) {
  return arr[arr.length - 1];
}

function prevTokenTypeCheck(tokens, type) {
  return tokens.length > 0 && tokens[tokens.length - 1].type === type;
}

export function tokenize(str) {
  let index = 0;
  let tokens = [];
  let indentStack = [0];
  let currentLine = 1;
  let currentCol = 1;
  const identifierRegex = /[#a-zA-Z0-9_\.]/;

  function identifierToken() {
    let char = next();
    let idStr = "";

    while (char !== undefined && identifierRegex.test(char)) {
      idStr += char;
      index += 1;
      char = next();
    }

    return idStr;
  }

  function commentToken() {
    let char = next();
    let comment = "";

    while (char !== undefined && char !== "\n" && char !== "\r") {
      comment += char;
      index += 1;
      char = next();
    }

    return comment;
  }

  function isWhitespace(char) {
    return /[\n\s\t]/.test(char);
  }

  function conditionToken() {
    let char = next();

    while (!identifierRegex.test(char)) {
      index += 1;
      char = next();
    }

    return identifierToken();
  }

  // this is the main function
  // It takes care of creating the right INDENT and DETENT tokens
  // the algorithm is taken from here - https://docs.python.org/3/reference/lexical_analysis.html
  // the implementation is mostly copied from the chevrotain example here - https://github.com/SAP/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js
  // TODO: don't add INDENT/DETENT tokens for lines which only have
  // whitespace
  function whitespaceTokenizer() {
    // the y ensures that this regex only matches the beginning of the string
    const regex = / +/y;
    let char = next();
    let wsCount = 0;

    // only checking for previous token as NEWLINE does not take
    // care of the first line
    if (prevTokenTypeCheck(tokens, "NEWLINE")) {
      const match = regex.exec(str.slice(index));
      let currentIndentLevel;
      if (match === null) {
        // this means that the new line does not have
        // any indentation. It's either empty or starts with a
        // non whitespace
        currentIndentLevel = 0;
      } else {
        currentIndentLevel = match[0].length;
      }

      const prevIndentLevel = last(indentStack);
      index += currentIndentLevel;
      currentCol = currentIndentLevel + 1;

      if (currentIndentLevel > prevIndentLevel) {
        indentStack.push(currentIndentLevel);
        return [
          {
            type: "INDENT",
            line: currentLine,
            col: 1,
            text: match[0]
          }
        ];
      } else if (currentIndentLevel < prevIndentLevel) {
        const dedentLevelInStack = indentStack.find(
          n => n === currentIndentLevel
        );

        // any dedent/outdent must match some previous indentation level.
        // otherwise it's a syntax error
        if (dedentLevelInStack === undefined) {
          throw new Error("Invalid indentation");
        }

        // keep popping indentation levels from indent dedentLevelInStack
        // until we reach the current indent level
        // push those many dedent tokens to tokenizer
        let indentLevelFromStack = last(indentStack);
        let dedentTokens = [];

        while (
          currentIndentLevel !== indentLevelFromStack &&
          indentStack.length > 0
        ) {
          indentStack.pop();
          dedentTokens.push({
            type: "DEDENT",
            line: currentLine,
            text: match ? match[0] : "",
            col: 1
          });

          indentLevelFromStack = last(indentStack);
        }

        return dedentTokens;
      } else {
        // same indentation level. do nothing. just consume it.
        return [];
      }
    } else {
      // TODO - should we separate this out into a whitespace tokenizer
      // and call this one indentDedentTokenizer?
      while (next() && /\t| /.test(next())) {
        index += 1;
      }

      return [];
    }
  }

  // return the next character in the stream
  function next() {
    return str[index];
  }

  function peek() {
    return str[index + 1];
  }

  function addToken(type, text) {
    tokens.push({
      type,
      text,
      line: currentLine,
      col: currentCol
    });

    currentCol += text ? text.length : 1;
  }

  while (index < str.length) {
    // after every round, let's just check if we need to
    // insert indent/dedent tokens
    tokens = tokens.concat(whitespaceTokenizer());
    const char = next();
    if (char === "\n") {
      addToken("NEWLINE");
      currentLine += 1;
      currentCol = 1;
      index += 1;
    } else if (char === "%") {
      const comment = commentToken();
      addToken("COMMENT", comment);
    } else if (char === "&") {
      addToken("PARALLEL_STATE");
      index += 1;
    } else if (char === "$") {
      addToken("FINAL_STATE");
      index += 1;
    } else if (char === "*") {
      addToken("INITIAL_STATE");
      index += 1;
    } else if (char === ";") {
      // we expect a condition after the semicolon
      const conditionName = conditionToken();
      addToken("CONDITION", conditionName);
    } else if (identifierRegex.test(char)) {
      const id = identifierToken();
      addToken("IDENTIFIER", id);
      // TODO: this check will not work when a dedent removes all
      // whitespace from a line. i.e. a line starts from the beginning
      // of the line
    } else if (char === " " || char === "\t") {
      const wsTokens = whitespaceTokenizer();
      tokens = tokens.concat(wsTokens);
    } else if (char === "-" && peek() === ">") {
      addToken("TRANSITION_ARROW");

      index += 2;
    } else if (char === ">") {
      index += 1;
      // we expect a condition after the semicolon
      const actionNames = conditionToken();
      addToken("ACTIONS", actionNames);
    } else {
      addToken("UNKNOWN", char);
      index += 1;
    }
  }

  // TODO - at the end of the tokenizing we need to pop out all remaining
  // indents from stack and push DEDENT tokens to our tokens list
  while (indentStack.length > 1 && indentStack.pop() > 0) {
    tokens.push({ type: "DEDENT", line: currentLine, col: currentCol });
  }
  return tokens;
}
