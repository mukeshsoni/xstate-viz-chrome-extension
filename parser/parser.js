import { tokenize } from "./tokenizer";

// TODO: What if instead of throwing exceptions, each of the parsers returned
// a Result type. The result type will encapsulate the possibility of failure

// omit certain properties from an Object
// the keys arguments contains array of strings which are
// the array of property names to be omitted
function omit(keys, obj) {
  return Object.entries(obj)
    .filter(([k, _]) => !keys.includes(k))
    .reduce((acc, [k, v]) => ({ ...acc, [k]: v }), {});
}

// merges array of objects into a single object
function arrayOfObjToObj(arr) {
  return arr.reduce((acc, item) => {
    // in case of transient states, we will have { '': { target: 'abc', cond: xyz } } kind of transitions. And they need to be merged for all '' appearances
    // They need to be merged into an array
    if (Object.keys(item).includes("")) {
      return {
        ...acc,
        "": acc[""] ? acc[""].concat(item[""]) : [item[""]]
      };
    } else {
      return { ...acc, ...item };
    }
  }, {});
}

function withInitialState(stateInfo) {
  const stateName = Object.keys(stateInfo)[0];
  const nestedStates = stateInfo[stateName].states;
  const nestedStateNames = Object.keys(nestedStates || {});

  if (nestedStateNames && nestedStateNames.length > 0) {
    const initialStateName = Object.entries(nestedStates).reduce(
      (acc, [k, v]) => {
        if (v.isInitial) {
          return k;
        } else {
          return acc;
        }
      },
      nestedStateNames[0]
    );

    return {
      ...stateInfo,
      [stateName]: {
        ...stateInfo[stateName],
        initial: initialStateName
      }
    };
  } else {
    return stateInfo;
  }
}

class ParserError extends Error {
  constructor(token, ...params) {
    super(...params);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ParserError);
    }

    this.token = token;
  }
}

// the main function. Just call this with the tokens
export function parse(inputStr) {
  const tokensToIgnore = ["COMMENT", "NEWLINE", "WS"];
  // 1. filter the comment tokens. Not useful for the parsing
  // 2. We can also treat newlines as useless. They were only useful during
  // the tokenizing phase because the INDENT and DEDENT tokens have be to be
  // preceded by a NEWLINE. In the final grammar, newlines only complicate things
  const tokens = tokenize(inputStr).filter(
    t => !tokensToIgnore.includes(t.type)
  );
  let index = 0;

  const consume = () => tokens[index++];

  // implements grammar rule with possibilities
  // using backtracking
  // e.g. operator -> '+' | '-' | '*' | '/'
  function oneOrAnother(...args) {
    const savedIndex = index;

    for (let i = 0; i < args.length; i++) {
      const parser = args[i];
      try {
        const parserResult = parser();
        return parserResult;
      } catch (e) {
        // else reset index
        index = savedIndex;
      }
    }

    // if none of the parsers worked
    throw new ParserError(
      tokens[index],
      `oneOrAnother parser: matched none of the rules: ${args
        .map(fn => fn.name)
        .join(" | ")}`
    );
  }

  function zeroOrOne(fn) {
    const savedIndex = index;

    try {
      const parserResult = fn();

      return [parserResult];
    } catch (e) {
      index = savedIndex;
      return [];
    }
  }

  // to implement things like statements = transitions * states*
  function zeroOrMore(fn) {
    const parserResults = [];

    while (true) {
      const savedIndex = index;

      try {
        const parserResult = fn();

        parserResults.push(parserResult);
      } catch (e) {
        index = savedIndex;
        return parserResults;
      }
    }
  }

  // for cases like A -> B+
  // where B can appear one or more times
  function oneOrMore(fn) {
    try {
      const parserResult = fn();

      return [parserResult].concat(zeroOrMore(fn));
    } catch (e) {
      return e;
    }
  }

  function identifier() {
    if (tokens[index].type === "IDENTIFIER") {
      return consume().text;
    }

    throw new ParserError(
      tokens[index],
      `Could not find IDENTIFIER. Instead found ${tokens[index]}`
    );
  }

  function actions() {
    if (tokens[index].type === "ACTIONS") {
      return consume().text;
    }

    throw new ParserError(
      tokens[index],
      `Could not find ACTIONS identifier. Instead found ${tokens[index]}`
    );
  }

  function condition() {
    if (tokens[index].type === "CONDITION") {
      return consume().text;
    }

    throw new ParserError(
      tokens[index],
      `Could not find CONDITION identifier. Instead found ${tokens[index]}`
    );
  }

  function parallelState() {
    if (consume().type === "PARALLEL_STATE") {
      return true;
    }

    throw new ParserError(tokens[index], "Expected PARALLEL_STATE");
  }

  function finalState() {
    if (consume().type === "FINAL_STATE") {
      return true;
    }

    throw new ParserError(tokens[index], "Expected PARALLEL_STATE");
  }

  function initialState() {
    if (consume().type === "INITIAL_STATE") {
      return true;
    }

    throw new ParserError(tokens[index], "Expected PARALLEL_STATE");
  }

  function indent() {
    if (consume().type === "INDENT") {
      return true;
    }

    throw new ParserError(tokens[index], "Expected indent");
  }

  function dedent() {
    if (consume().type === "DEDENT") {
      return true;
    }

    throw new ParserError(tokens[index], "Expected dedent");
  }

  function arrow() {
    if (consume().type === "TRANSITION_ARROW") {
      return true;
    }

    throw new ParserError(tokens[index], "expected arrow");
  }

  function transition() {
    // we might have arrow as start for transient states
    const eventName = zeroOrOne(identifier) || "";
    arrow();
    const stateName = identifier();
    let conditionName;
    let actionNames;

    if (eventName) {
      conditionName = zeroOrOne(condition);
      actionNames = zeroOrMore(actions);
    } else {
      // if the first event name was absent, the condition is mandatory
      conditionName = condition();
      actionNames = zeroOrMore(actions);
    }

    return {
      type: "transition",
      // TODO: What if we used the more verbose definition of each transition
      // from the parser. { x: { target: 'y', cond: 'abc' }}. If we want to do
      // any optimizations, like convert transitions without any conditions to
      // shorter form, like { x: 'y' }, it can be done later on in one fell
      // swoop
      [eventName]:
        conditionName.length > 0 || actionNames.length > 0
          ? {
              target: stateName,
              cond: conditionName.length > 0 ? conditionName[0] : undefined,
              actions: actionNames.length > 0 ? actionNames : undefined
            }
          : stateName
    };
  }

  function stateWithNameOnly() {
    const stateName = identifier();
    const parallel = zeroOrOne(parallelState);
    const isFinal = zeroOrOne(finalState);
    const isInitial = zeroOrOne(initialState);

    return {
      [stateName]: {
        type:
          parallel.length > 0
            ? "parallel"
            : isFinal.length > 0
            ? "final"
            : undefined,
        isInitial: isInitial.length > 0 ? true : undefined
      }
    };
  }
  // like transitions, nested states etc.
  // e.g.
  // active
  //  click_checkbox -> active
  //  uncheck -> inactive
  function stateWithMoreDetails() {
    const stateName = identifier();
    const parallel = zeroOrOne(parallelState);
    const isFinal = zeroOrOne(finalState);
    const isInitial = zeroOrOne(initialState);
    const isIndentThere = zeroOrOne(indent);
    let transitionsAndStates = [];

    // if there is an indent after state name, it has to be state with
    // extra info
    if (isIndentThere.length > 0) {
      transitionsAndStates = oneOrMore(() => {
        return oneOrAnother(transition, stateParser);
      });

      // any rule which has an indent should be always accompanied by a closing
      // dedent. The indent and dedent have to match up, just like parentheses
      // in other languages.
      dedent();
    }

    const transitions = transitionsAndStates.filter(
      ts => ts.type === "transition"
    );
    const nestedStates = transitionsAndStates.filter(
      ts => ts.type !== "transition"
    );

    return {
      [stateName]: {
        type:
          parallel.length > 0
            ? "parallel"
            : isFinal.length > 0
            ? "final"
            : undefined,
        isInitial: isInitial.length > 0 ? true : undefined,
        on:
          transitions.length > 0
            ? omit(["type"], arrayOfObjToObj(transitions))
            : undefined,
        states:
          nestedStates.length > 0 ? arrayOfObjToObj(nestedStates) : undefined
      }
    };
  }

  function stateParser() {
    try {
      const stateInfo = stateWithMoreDetails();

      return withInitialState(stateInfo);
    } catch (e) {
      throw new ParserError(tokens[index], e);
    }
  }

  function stateMachine() {
    try {
      const parserOutput = stateParser();

      const id = Object.keys(parserOutput)[0];

      let initial = undefined;

      if (parserOutput[id].states) {
        initial = Object.keys(parserOutput[id].states)[0];
      }

      return {
        id,
        initial,
        ...parserOutput[id]
      };
    } catch (e) {
      return { error: e };
    }
  }

  return stateMachine();
}
