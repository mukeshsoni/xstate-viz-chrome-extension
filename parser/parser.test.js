import { tokenize } from "./tokenizer";
import { parse } from "./parser";

const inputStr = `abc
% some comment
  def -> lmn
  pasta -> noodles %more comment
  ast&*
    opq -> rst; ifyes
    uvw -> #abc.lastState
    nestedstate1
    nestedstate2*
  tried -> that > andDoThis
  lastState
    % trying out transient state
    -> ast; ifyes
    -> lastState; ifno`;

const expectedXstateJSON = {
  id: "abc",
  initial: "ast",
  on: {
    def: "lmn",
    pasta: "noodles",
    tried: {
      target: "that",
      actions: ["andDoThis"]
    }
  },
  states: {
    ast: {
      type: "parallel",
      initial: "nestedstate2",
      isInitial: true,
      on: {
        opq: { target: "rst", cond: "ifyes" },
        uvw: "#abc.lastState"
      },
      states: {
        nestedstate1: {},
        nestedstate2: { isInitial: true }
      }
    },
    lastState: {
      on: {
        "": [
          {
            target: "ast",
            cond: "ifyes"
          },
          {
            target: "lastState",
            cond: "ifno"
          }
        ]
      }
    }
  }
};

const invalidInputStr = `abc
  def -> lmn
      pqr
    stm`;

const fetchInputStr = `fetch
  idle
      FETCH -> loading
  loading
      RESOLVE -> success
      REJECT -> failure
  success$
  failure
      RETRY -> loading`;

const expectedXstateJSONFetch = {
  id: "fetch",
  initial: "idle",
  states: {
    idle: {
      on: {
        FETCH: "loading"
      }
    },
    loading: {
      on: {
        RESOLVE: "success",
        REJECT: "failure"
      }
    },
    success: {
      type: "final"
    },
    failure: {
      on: {
        RETRY: "loading"
      }
    }
  }
};
describe("tokenizer", () => {
  it("should give the correct number of tokens", () => {
    const tokens = tokenize(inputStr);

    console.log(tokens);
    expect(tokens).toHaveLength(53);
  });

  it("gives correct indent and dedent tokens", () => {
    const tokens = tokenize(inputStr);

    expect(tokens[2].type).toEqual("COMMENT");
    expect(tokens[4].type).toEqual("INDENT");
    expect(tokens[18].type).toEqual("INDENT");
    expect(tokens[33].type).toEqual("DEDENT");
  });

  it("catches incorrect indentation errors", () => {
    expect(() => tokenize(invalidInputStr)).toThrowError("Invalid indentation");
  });

  it("should have line and column number for tokens", () => {
    const tokens = tokenize(inputStr);

    const secondIdentifier = tokens[5];

    expect(secondIdentifier.type).toEqual("IDENTIFIER");
    expect(secondIdentifier.line).toEqual(3);
    expect(secondIdentifier.col).toEqual(3);

    const lastToken = tokens[tokens.length - 1];
    expect(lastToken.type).toEqual("DEDENT");

    const thirdLastToken = tokens[tokens.length - 4];

    expect(thirdLastToken.type).toEqual("IDENTIFIER");
    expect(thirdLastToken.line).toEqual(14);
    expect(thirdLastToken.col).toEqual(6);

    const uvwToken = tokens.find(token => token.text === "uvw");

    expect(uvwToken.type).toEqual("IDENTIFIER");
    expect(uvwToken.line).toEqual(7);
    expect(uvwToken.col).toEqual(5);
  });

  it("should have dedent after nestedstate2", () => {
    const tokens = tokenize(inputStr);

    const nestedstate2TokenIndex = tokens.findIndex(
      t => t.type === "IDENTIFIER" && t.text === "nestedstate2"
    );

    expect(tokens[nestedstate2TokenIndex + 3].type).toEqual("DEDENT");
  });
});

describe("parser", () => {
  it("should generate xstate representation of the input string", () => {
    const ast = parse(inputStr);

    // console.log(JSON.stringify(ast, null, 2));
    expect(ast).toEqual(expectedXstateJSON);
  });

  it("should generate xstate representation for the fetch statechart", () => {
    const ast = parse(fetchInputStr);

    // console.log(JSON.stringify(ast, null, 2));
    expect(ast).toEqual(expectedXstateJSONFetch);
  });

  it("should throw error when transition state not specified", () => {
    // TODO: Without the newline check in the parser
    // the below text is syntatically correct :(
    const inputStr = `abc
  def
    lmn ->
    -> lrt`;

    // TODO: The below string is also accepted with current grammar
    // This is because the rule
    // oneOrAnother(stateWithMoreDetails, stateWithNameOnly);
    // is also not correct. If a statename is followed by an indent, it should
    // try to resolve a stateWithMoreDetails. Not backtrack and try state
    // with name only
    // const inputStr = `abc
    // def
    // lmn ->
    // -> lrt`;
    const ast = parse(inputStr);

    console.log(ast.error.message, "\n", ast.error.token);
    console.log(JSON.stringify(ast, null, 2));
    expect(ast.error).toBeDefined();
  });
});

// tests cases when the parser finds and error in the input
// We want to give errors as accurate as possible
// And even helpful, if possible
describe("parser error handling", () => {});
