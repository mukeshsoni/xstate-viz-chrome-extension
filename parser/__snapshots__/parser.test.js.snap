// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should generate xstate representation for the fetch statechart 1`] = `
Object {
  "id": "fetch",
  "initial": "idle",
  "isInitial": undefined,
  "on": undefined,
  "states": Object {
    "failure": Object {
      "isInitial": undefined,
      "on": Object {
        "RETRY": "loading",
      },
      "states": undefined,
      "type": undefined,
    },
    "idle": Object {
      "isInitial": undefined,
      "on": Object {
        "FETCH": "loading",
      },
      "states": undefined,
      "type": undefined,
    },
    "loading": Object {
      "isInitial": undefined,
      "on": Object {
        "REJECT": "failure",
        "RESOLVE": "success",
      },
      "states": undefined,
      "type": undefined,
    },
    "success": Object {
      "isInitial": undefined,
      "on": undefined,
      "states": undefined,
      "type": "final",
    },
  },
  "type": undefined,
}
`;

exports[`tokenizer should give the correct number of tokens 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "text": "abc",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 4,
    "line": 1,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 1,
    "line": 2,
    "text": "% some comment",
    "type": "COMMENT",
  },
  Object {
    "col": 15,
    "line": 2,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 1,
    "line": 3,
    "text": "  ",
    "type": "INDENT",
  },
  Object {
    "col": 3,
    "line": 3,
    "text": "def",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 6,
    "line": 3,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 7,
    "line": 3,
    "text": "lmn",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 10,
    "line": 3,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 3,
    "line": 4,
    "text": "pasta",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 8,
    "line": 4,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 9,
    "line": 4,
    "text": "noodles",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 16,
    "line": 4,
    "text": "%more comment",
    "type": "COMMENT",
  },
  Object {
    "col": 29,
    "line": 4,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 3,
    "line": 5,
    "text": "ast",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 6,
    "line": 5,
    "text": undefined,
    "type": "PARALLEL_STATE",
  },
  Object {
    "col": 7,
    "line": 5,
    "text": undefined,
    "type": "INITIAL_STATE",
  },
  Object {
    "col": 8,
    "line": 5,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 1,
    "line": 6,
    "text": "    ",
    "type": "INDENT",
  },
  Object {
    "col": 5,
    "line": 6,
    "text": "opq",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 8,
    "line": 6,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 9,
    "line": 6,
    "text": "rst",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 12,
    "line": 6,
    "text": "ifyes",
    "type": "CONDITION",
  },
  Object {
    "col": 17,
    "line": 6,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 5,
    "line": 7,
    "text": "uvw",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 8,
    "line": 7,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 9,
    "line": 7,
    "text": "#abc.lastState",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 23,
    "line": 7,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 5,
    "line": 8,
    "text": "nestedstate1",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 17,
    "line": 8,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 5,
    "line": 9,
    "text": "nestedstate2",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 17,
    "line": 9,
    "text": undefined,
    "type": "INITIAL_STATE",
  },
  Object {
    "col": 18,
    "line": 9,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 1,
    "line": 10,
    "text": "  ",
    "type": "DEDENT",
  },
  Object {
    "col": 3,
    "line": 10,
    "text": "tried",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 8,
    "line": 10,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 9,
    "line": 10,
    "text": "that",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 13,
    "line": 10,
    "text": "andDoThis",
    "type": "ACTIONS",
  },
  Object {
    "col": 22,
    "line": 10,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 3,
    "line": 11,
    "text": "lastState",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 12,
    "line": 11,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 1,
    "line": 12,
    "text": "    ",
    "type": "INDENT",
  },
  Object {
    "col": 5,
    "line": 12,
    "text": "% trying out transient state",
    "type": "COMMENT",
  },
  Object {
    "col": 33,
    "line": 12,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 5,
    "line": 13,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 6,
    "line": 13,
    "text": "ast",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 9,
    "line": 13,
    "text": "ifyes",
    "type": "CONDITION",
  },
  Object {
    "col": 14,
    "line": 13,
    "text": undefined,
    "type": "NEWLINE",
  },
  Object {
    "col": 5,
    "line": 14,
    "text": undefined,
    "type": "TRANSITION_ARROW",
  },
  Object {
    "col": 6,
    "line": 14,
    "text": "lastState",
    "type": "IDENTIFIER",
  },
  Object {
    "col": 15,
    "line": 14,
    "text": "ifno",
    "type": "CONDITION",
  },
  Object {
    "col": 19,
    "line": 14,
    "type": "DEDENT",
  },
  Object {
    "col": 19,
    "line": 14,
    "type": "DEDENT",
  },
]
`;
